File: C/nQueen.c
#include <stdio.h>
#include <stdbool.h>

// Function to print the chessboard solution
void printSolution(int N, int board[N][N]) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++)
            printf("%2d ", board[i][j]);
        printf("\n");
    }
}

// Function to check if it's safe to place a queen
bool isSafe(int N, int board[N][N], int row, int col) {
    // Check this row on the left side
    for (int i = 0; i < col; i++)
        if (board[row][i])
            return false;

    // Check upper diagonal on the left side
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--)
        if (board[i][j])
            return false;

    // Check lower diagonal on the left side
    for (int i = row, j = col; i < N && j >= 0; i++, j--)
        if (board[i][j])
            return false;

    return true;
}

// Utility function to solve N-Queens problem using backtracking
bool solveNQueensUtil(int N, int board[N][N], int col) {
    if (col >= N)
        return true;

    for (int i = 0; i < N; i++) {
        if (isSafe(N, board, i, col)) {
            board[i][col] = 1; // Place queen

            // Recur to place the rest of the queens
            if (solveNQueensUtil(N, board, col + 1))
                return true;

            board[i][col] = 0; // Backtrack
        }
    }

    return false;
}

// Function to solve the N-Queens problem
bool solveNQueens(int N) {
    int board[N][N];

    // Initialize the board
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            board[i][j] = 0;

    // Start solving the problem
    if (!solveNQueensUtil(N, board, 0)) {
        printf("Solution does not exist.\n");
        return false;
    }

    printf("Solution for %d-Queens problem:\n", N);
    printSolution(N, board);
    return true;
}

// Main function to execute the program
int main() {
    int N;

    printf("Enter the value of N: ");
    scanf("%d", &N);

    solveNQueens(N);

    return 0;
}

File: Python/N_Queen.py
def solveNQueens(N: int):
    def is_safe(board, row, col):
        # Check this column on upper side
        for i in range(row):
            if board[i][col] == 'Q':
                return False

        # Check upper diagonal on left side
        for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
            if board[i][j] == 'Q':
                return False

        # Check upper diagonal on right side
        for i, j in zip(range(row, -1, -1), range(col, N)):
            if board[i][j] == 'Q':
                return False

        return True

    def solve(board, row):
        if row == N:
            # Store the configuration
            solution = ["".join(row) for row in board]
            solutions.append(solution)
            return

        for col in range(N):
            if is_safe(board, row, col):
                board[row][col] = 'Q'
                solve(board, row + 1)
                board[row][col] = '.'

    # Initialize the chessboard
    board = [['.' for _ in range(N)] for _ in range(N)]
    solutions = []
    solve(board, 0)
    return solutions


def print_solution(solution):
    board_size = len(solution)
    print("\n" + "=" * (2 * board_size - 1))
    for row in solution:
        print(" | ".join(row))
        print("-" * (2 * board_size - 1))
    print("=" * (2 * board_size - 1) + "\n")


def main():
    # Get N value from the user
    try:
        N = int(input("Enter the value of N (between 1 and 10): "))
    except ValueError:
        print("Please enter a valid integer.")
        return

    # Validate the input
    if N <= 0 or N > 10:
        print("Enter N value from 1 to 10.")
        return

    # Solve N-Queen problem
    solutions = solveNQueens(N)

    if solutions:
        for idx, solution in enumerate(solutions, start=1):
            print(f"Solution {idx}:")
            print_solution(solution)
    else:
        print("No solution exists for N =", N)


# Run the program
if __name__ == "__main__":
    main()
